python_modules/dagster/docs/reference.rst

```diff
@@ -0,0 +1,212 @@
+
+Materializations make it possible to introspect the intermediate state of a pipeline execution
+and ask questions like, "Exactly what output did this solid have on this particular run?" This is
+useful when auditing or debugging pipelines, and makes it possible to establish the `provenance` of
+data artifacts.
+
+Materializations also enable partial re-execution of pipelines "starting from" a materialized state
+of the upstream execution. This is useful when a pipeline fails halfway through, or in order to
+explore how new logic in part of a pipeline would have operated on outputs from previous runs of
+the pipeline.
+
+Expectation
+^^^^^^^^^^^
+
+An expectation is a predicate on a solid’s output. Expectations can be used to enforce runtime data
+quality and integrity constraints, so that pipelines fail early -- before any downstream solids
+execute on bad data.
+
+Expectations are defined using the :class:`ExpectationDefinition <dagster.ExpectationDefinition>`
+class. We also provide a `thin wrapper <https://github.com/dagster-io/dagster/tree/master/python_modules/libraries/dagster-ge>`_
+around the `great_expectations <https://github.com/great-expectations/great_expectations>`_ library
+so you can use its existing repertoire of expectartions with Dagster.
+
+.. _pipeline:
+
+Pipeline
+^^^^^^^^
+Data pipelines are directed acyclic graphs (DAGs) of solids -- that is, they are made up of a number
+of solids which have data `dependencies <#dependency-definition>`__ on each other (but no circular
+dependencies), along with a set of associated pipeline context definitions, which declare the various
+environments in which a pipeline can execute.
+
+Pipelines are defined using the :class:`PipelineDefinition <dagster.PipelineDefinition>` class, and
+their contexts are defined using :class:`PipelineContextDefinition <dagster.PipelineContextDefinition>`.
+
+When a pipeline is combined with a given config conforming to one of its declared contexts, it can
+be compiled by the Dagster engine into an execution plan that can be executed on various compute
+substrates.
+
+Concretely, a pipeline might include context definitions for local testing (where databases and
+other resources will be mocked, in-memory, or local) and for running in production (where resources
+will require different credentials and expose configuration options). When a pipeline is compiled
+with a config corresponding to one of these contexts, it yields an execution plan suitable for the
+given environment.
+
+Repository
+^^^^^^^^^^
+
+A repository is a collection of pipelines that can be made available to the Dagit UI and other
+higher-level tools. Repositories are defined using the
+:class:`RepositoryDefinition <dagster.RepositoryDefinition>` class, and made available to
+higher-level tools with a special ``repository.yml`` file that tells the tools where to look for a
+repository definition.
+
+Config
+^^^^^^
```

Comment:
should this be Environment Config?

Reply:
I'm not sure about this, we have had so many different words for this thing. @schrockn ?